### PR TITLE
Less punishing punishment

### PR DESCRIPTION
**Pull request by PoweredByPot**
minor tweaks to CSATpunish to hopefully stop the instant lose, and penalizing rebels almost always.

### CODE REVIEW

File: A3-Antistasi/statSave/saveLoop.sqf [hunk 1]
```diff
@@ -208,7 +208,7 @@ if (!isDedicated) then
 			_typesX pushBackUnique _x;
 			};
 		};
-	} forEach ["AS","CON","DES","LOG","RES","CONVOY","DEF_HQ","AttackAAF"];
+	} forEach ["AS","CON","DES","LOG","RES","CONVOY","DEF_HQ","rebelAttack","invaderPunish"];
```
Comment (Spoffy) [inline A3-Antistasi/statSave/saveLoop.sqf]: Punishment missions won't be restored properly, as saveFuncs.sqf doesn't know how to restore them

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf [hunk 2]
```diff
@@ -1,60 +1,60 @@
...
+private _sideTarget = if (sidesX getVariable [_attackDestination,sideUnknown] == Occupants) then {Occupants} else {teamPlayer};
+_missionExpireTime = time + 3600;
 
-private _CSATTransportAir = vehCSATTransportHelis + vehCSATTransportPlanes;
+private _invaderAirTransport = vehCSATTransportHelis + vehCSATTransportPlanes;
 
 for "_i" from 1 to 3 do
 	{
-	_typeVehX = if (_i != 3) then {selectRandom (vehCSATAir select {[_x] call A3A_fnc_vehAvailable})} else {selectRandom (_CSATTransportAir select {[_x] call A3A_fnc_vehAvailable})};
+	_attackVehicle = if (_i != 3) then {selectRandom (vehCSATAir select {[_x] call A3A_fnc_vehAvailable})} else {selectRandom (_invaderAirTransport select {[_x] call A3A_fnc_vehAvailable})};
```
Comment (Spoffy) [inline A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: This rename from '_typeVehX' to 'attackVehicle' isn't a good rename - because we no longer know it's the *type* of a vehicle (i.e, the class name). It's also not necessarily an attack vehicle - it could be a transport. 
I suggest '_typeAirVehicle' instead.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf [hunk 3]
```diff
@@ -109,96 +109,98 @@ for "_i" from 1 to 3 do
...
 
-_civilMax = {alive _x} count _civiles;
-_solMax = count _soldiers;
+_soldiersSpawned = count _soldiers;
 
-for "_i" from 0 to round random 2 do
+if (tierWar >= 5) then
 	{
-	if ([vehCSATPlane] call A3A_fnc_vehAvailable) then
+	for "_i" from 0 to floor random 1 do
```
Comment (Spoffy) [inline A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: This will only ever loop once.
'random 1' returns a value between 0 and 1. The odds of it having *exactly* the value 1 (which is the only value which would cause it to loop twice) are astronomically small.

Given you've added a tierWar requirement, I suggest removing the random and having it be 2 loops.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf [hunk 4]
```diff
@@ -109,96 +109,98 @@ for "_i" from 1 to 3 do
...
-	if ({(side _x == teamPlayer) and (_x distance _posDestination < _size * 2)} count allUnits >= {(side _x == Occupants) and (_x distance _posDestination < _size * 2)} count allUnits) then
+	//["invaderPunish", "SUCCEEDED",true] spawn BIS_fnc_taskSetState;
+	["invaderPunish",[format ["%2 is attacking innocent civilians in %1! Defend the city at all costs",_nameDestination,nameInvaders],format ["%1 Punishment",nameInvaders],_attackDestination],getMarkerPos _attackDestination,"SUCCEEDED"] call A3A_fnc_taskUpdate;
+	if ({(side _x == teamPlayer) and (_x distance _posDestination < _size * 2)} count allUnits >= {(side _x == _sideTarget) and (_x distance _posDestination < _size * 2)} count allUnits) then
 		{
-		if (sidesX getVariable [_mrkDestination,sideUnknown] == Occupants) then {[-15,15,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]} else {[-5,15,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]};
+		if (sidesX getVariable [_attackDestination,sideUnknown] == Occupants) then {[-15,15,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]} else {[-5,15,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]};
```
Comment (Spoffy) [inline A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: Reading this through - it doesn't actually make any sense to change the city's support here. It should really reduce the support of both factions. This is unrelated to your changes.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf [hunk 5]
```diff
@@ -109,96 +109,98 @@ for "_i" from 1 to 3 do
...
 		[-5,0] remoteExec ["A3A_fnc_prestige",2];
 		{[-10,10,_x] remoteExec ["A3A_fnc_citySupportChange",2]} forEach citiesX;
 		{if (isPlayer _x) then {[10,_x] call A3A_fnc_playerScoreAdd}} forEach ([500,0,_posDestination,teamPlayer] call A3A_fnc_distanceUnits);
 		[10,theBoss] call A3A_fnc_playerScoreAdd;
 		}
 	else
 		{
-		if (sidesX getVariable [_mrkDestination,sideUnknown] == Occupants) then {[15,-5,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]} else {[15,-15,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]};
-		{[10,-10,_x] remoteExec ["A3A_fnc_citySupportChange",2]} forEach citiesX;
+		if (sidesX getVariable [_attackDestination,sideUnknown] == Occupants) then {[15,0,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]} else {[15,0,_posDestination] remoteExec ["A3A_fnc_citySupportChange",2]};
```
Comment (Spoffy) [inline A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: If we've succeeded, but the owners of the town outnumber players, we should give the owners of the town some credit, instead of the opposite faction.

File: A3-Antistasi/statSave/saveLoop.sqf (hunk 1 above)
Comment (PoweredByPot) [reply A3-Antistasi/statSave/saveLoop.sqf]: I need to find a better solution here then. As it existed, when trying to restore the mission, what was an attack on a rebel outpost could become a CSAT punishment mission, even though the conditions were otherwise met.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 3 above)
Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: I ran into this issue when deciding the method for loot crate parameters.

The 'random' command will return anything up to X.99, so in this case up to 1.9. We used to use round instead of floor, and that mean that '0 to round random 1' would mean potentially up to 3.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 4 above)
Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: I don't know, it seems a great opportunity to prove you are "the hero of the people", and it seems logical it would spread to other towns. I don't know about the whole country though.

File: A3-Antistasi/statSave/saveLoop.sqf (hunk 1 above)
Comment (Spoffy) [reply A3-Antistasi/statSave/saveLoop.sqf]: You just need to add logic to saveFuncs.sqf to restore the punishment mission. You'll have to do that, no matter the solution you pick.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 3 above)
Comment (Spoffy) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: No, random returns a value between 0 and X. Test it yourself - put 'random 1' in the debug console and spam execute it.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 4 above)
Comment (Spoffy) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: Right - but neither of these improve support for the player. They either decrease Occupant and increase Invader, or vice-versa.

Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: incorrect sir. there is only rebel and occupant support.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 3 above)
Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: fixed

File: A3-Antistasi/statSave/saveLoop.sqf (hunk 1 above)
Comment (PoweredByPot) [reply A3-Antistasi/statSave/saveLoop.sqf]: wait, this was intentional.

Because I don't know that all of this will necessarily fix the mission failing right away, I would rather the task NOT be restored on load, so that a city isn't wiped off the map just because the server restarted.

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 2 above)
Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: fixed

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 5 above)
Comment (PoweredByPot) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: the win condition says "if rebels are equal to or outnumber owners (either rebels or occupants) then {rebel reward small occupant penalty} else {occupant reward and no change to rebels}.

The thought here was:

1) if the town is held by occupants and they do a better job defending it then give them a boost, but don't penalize the rebels.

2) if the town is held by occupants but REBELS do a better job defending then give rebels a big boost and occupants a big penalty.

3) If the town is held by rebels, then occupants don't even show up so they can only be the dominant faction. If they win, another big boost

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 4 above)
Comment (Spoffy) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: Sorry, you are perfectly right - I've been thrown by objectively wrong naming conventions!

File: A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf (hunk 5 above)
Comment (Spoffy) [reply A3-Antistasi/functions/CREATE/fn_invaderPunish.sqf]: In light of how citySupportChange actually works, you're right.

File: A3-Antistasi/statSave/saveLoop.sqf (hunk 1 above)
Comment (Spoffy) [reply A3-Antistasi/statSave/saveLoop.sqf]: Alright, I've opened issue #344 so we won't forget.
